Allow opting into analytics

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

#### Changes
* Copies BMO's analytics work from Genesis Blocks into GCB
* Port the Settings page PHP from Pro to free
* Not in scope: using analytics, like sending an event

Fixes [GF-3267](https://wpengine.atlassian.net/browse/GF-3267)

#### Testing instructions
1. `composer i && npm i && npm run dev`
2. /wp-admin > Custom Blocks > Settings
3. Check 'Enable analytics':
![Screen Shot 2021-12-16 at 10 26 01 AM](https://user-images.githubusercontent.com/4063887/146409946-cb5cad1e-8d59-4959-bda9-e241fafd2d52.png)
4. Click 'Save Changes'
5. Create a new post
6. In the DevTools, enter `window.GcbAnalytics`
7. Expected: It's defined: 
![Screen Shot 2021-12-16 at 10 26 50 AM](https://user-images.githubusercontent.com/4063887/146410112-5b151e1c-f54f-4ba8-b6a5-58fe9f4d4f7d.png)

kienstra committed:
Change node version to 14.18.1, see if that makes builds pass

kienstra committed:
Mainly copy-paste the Settings page PHP to the free plugin

kienstra committed:
Correct 2 package tags to be for free, not pro

kienstra committed:
Change .nvmrc to 14, as lts is now 16

It looks like WP is still
on node version 14.

kienstra committed:
Move the settings registration to the top of Admin class

kienstra committed:
Change the settings group name to align with what it does

kienstra committed:
Add a Settings.php view, copied from the Pro plugin

Though it only has free plugin logic,
wiht an action for the Pro plugin (or anything else)
to run.

kienstra committed:
Enqueue the analytics script if opted-in, props BMO

Thanks to BMO for:
https://github.com/studiopress/genesis-blocks/blob/81b39c336123748bda7e4b2724cf9dbcea13c48b/lib/Analytics/AssetManager.php#L71

kienstra committed:
Add a test for having opted into analytics

kienstra commented on php/Blocks/Loader.php:
```diff
@@ -158,6 +166,16 @@ public function editor_assets() {
+		if ( Settings::ANALYTICS_OPTED_IN_VALUE === get_option( Settings::ANALYTICS_OPTION_NAME ) ) {
+			wp_enqueue_script(
+				self::ANALYTICS_SCRIPT_SLUG,
+				'https://www.googletagmanager.com/gtag/js?id=UA-17364082-14',
```
The UA id should probably be unique to GCB

BMO-tech replied:
Should be able to get that from the GA dashboard

kienstra replied:
OK nice, could you Slack me a link to that?

kienstra committed:
Bump copyright dates that I should have bumped earlier

kienstra committed:
Copy BMO's wp_localize_script() call

https://github.com/studiopress/genesis-blocks/blob/81b39c336123748bda7e4b2724cf9dbcea13c48b/lib/Analytics/AssetManager.php#L79

kienstra committed:
Copy BMO's GAClient into GCB

https://github.com/studiopress/genesis-blocks/blob/81b39c336123748bda7e4b2724cf9dbcea13c48b/lib/Analytics/js/gaClient.js

kienstra committed:
Instantiate GAClient in the GCB editor and the block editor

kienstra committed:
Remove the GB Analytics ID, add Todo to replace with a unique one for GCB

kienstra committed:
Change window.GenesisAnalytics to not conflict with Genesis Blocks

kienstra committed:
Move the analytics bootstrapping inside domReady()

kienstra commented on .nvmrc:
```diff
@@ -1 +1 @@
-lts/*
+14
```
Node LTS used to be `14`, but now it's `16`. It seems that Gutenberg [still uses](https://github.com/WordPress/gutenberg/blob/408bcd819fb60cfffe7fa10f57c86be0d5693ec6/.nvmrc) `14`

johnstonphilip replied:
interesting! I haven't heard of this `lts` before.

kienstra commented on js/src/common/classes/index.js:
```diff
@@ -0,0 +1 @@
+export { default as GAClient } from './GAClient';
```
This is silly for one file, but it follows the pattern of exporting files in all of the other directories

kienstra committed:
Acknowledge BMO's work in debounce()

kienstra committed:
Change some copypasta: this was copied from a Subscription page

kienstra committed:
Acknowledge BMO's work in the GAClient

kienstra committed:
Make comment format more like GCB's comments

BMO-tech commented on js/src/common/classes/GAClient.js:
```diff
@@ -0,0 +1,100 @@
+		};
+
+		// @ts-ignore
+		this.config = window.genesisAnalyticsConfig || {};
```
Does this value need to be `window.gcbAnalyticsConfig` ?

kienstra replied:
Uf, you're right.

kienstra replied:
Fixed in https://github.com/studiopress/genesis-custom-blocks/pull/101/commits/b45e3a3b77e332e894a49df17ef24da30794203e

BMO-tech commented on js/src/block-editor/index.js:
```diff
@@ -3,15 +3,29 @@
+// @ts-ignore
 registerBlocks( genesisCustomBlocks, gcbBlocks, Edit );
+
+domReady( () => {
```
`domReady` is interesting. Where you coming across any issues that needed this?

kienstra replied:
That's a good point, I didn't notice that it was needed.

Moving it out of `domReady()` now…

kienstra replied:
Moved out in https://github.com/studiopress/genesis-custom-blocks/pull/101/commits/87d1f456f42c97069065698e45448101c43215b7

BMO-tech commented on js/src/block-editor/index.js:
```diff
@@ -3,15 +3,29 @@
+
+domReady( () => {
+	// @ts-ignore
+	window.dataLayer = window.dataLayer || [];
```
What was you logic for moving `window.dataLayer` our of `GAClient` ?

kienstra replied:
Not a functional reason, I just like class files to have only the class, and no logic other than that.

Maybe from too much time in PHP 🤣 

But if you'd prefer, I could move it back.

kienstra replied:
Actually, I'm going to move it back.

kienstra replied:
Moved it back in https://github.com/studiopress/genesis-custom-blocks/pull/101/commits/9d8bd80bc6dc9781f6f2fdb8f24ece00ed3408d3

BMO-tech commented on js/src/common/helpers/debounce.js:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @return {() => void} A debounced function.
+ */
+const debounce = ( func, wait ) => {
```
It doesn't follow the same standard as the rest of this plugin, but it is possible to name the export instead of using `default`.

```
export const debounce = () => { ... }
```

That will prevent the need for `export default debounce` at the bottom, and it can be included with

```
include { debounce }  from './debounce';
```

kienstra replied:
Yeah, this plugin's setup is a bit outdated, with the `index.js` file having all of the exports.

Ideally, I'd modernize the whole `js/src` directory.

But it might be strange to not have this `debounce` in `index.js`, while all of the other helper functions are in it.

BMO-tech commented on php/Blocks/Loader.php:
```diff
@@ -158,6 +166,22 @@ public function editor_assets() {
 			]
 		);
 
+		if ( Settings::ANALYTICS_OPTED_IN_VALUE === get_option( Settings::ANALYTICS_OPTION_NAME ) ) {
```
Do you want to factor in any historical opt-in events like we just merged into GB?

kienstra replied:
Hm, good question. 

That might be something for product, but I think for this scope, this might be enough.

kienstra committed:
Change genesisAnalyticsConfig to gcbAnalyticsConfig

There is no genesisAnalyticsConfig
in this plugin.

kienstra committed:
Change genesisAnalyticsConfig to gcbAnalyticsConfig

There is no genesisAnalyticsConfig
in this plugin.

kienstra committed:
Move window.dataLayer assignment back into the class file

It's an essential part of the class,
so it's strange to not have it in that file.

johnstonphilip commented on js/src/common/classes/GAClient.js:
```diff
@@ -0,0 +1,103 @@
+	 * @param {boolean | number | string} enable The value to be set.
+	 */
+	enableAnalytics( enable ) {
+		enable = !! +enable;
```
I'm not sure I'm familiar with this syntax. What is happening on this line?

kienstra replied:
Ah, I stole [that](https://github.com/studiopress/genesis-blocks/blob/c468ce9c049ad6d0ae260f698f445d5e0c6a9b1a/lib/Analytics/js/gaClient.js#L86) from BMO. 

I didn't know about `+` either.

It's first casting `enable` to a `Number` with `+enable`. For example, converting `'0'` to `0`.

Then, it casts it to a `Boolean` with `!!`.

I think this prevents an `enable` value of `'0'` from being `true`.

Like:

```js
if ( '0' ) {
    console.log( 'This is true' );
}
```

johnstonphilip replied:
Thanks!

kienstra replied:
Of course! I just found out about `+`

johnstonphilip commented:
Looks all good here!
<img width="584" alt="Screen Shot 2021-12-20 at 1 12 55 PM" src="https://user-images.githubusercontent.com/7538525/146813802-5475189d-e9d5-46e5-ab29-5b1cff185ac5.png">
<img width="564" alt="Screen Shot 2021-12-20 at 1 13 02 PM" src="https://user-images.githubusercontent.com/7538525/146813804-4cca38b4-7150-4292-aca4-51d686b0d120.png">
<img width="214" alt="Screen Shot 2021-12-20 at 1 14 19 PM" src="https://user-images.githubusercontent.com/7538525/146813805-15adabc0-67bc-4bf0-86e4-d7709b77dc0d.png">

kienstra commented:
Hi @johnstonphilip,
Thanks a lot for reviewing and testing this!

BMO-tech reviewed (approved):
Great work!

kienstra commented:
@BMO-tech, thanks so much for your testing and review.
